tinygrad/codegen/uopgraph.py

```diff
@@ -477,9 +481,28 @@ def delete_redundant_gates(root:UOp) -> Optional[UOp]:
   def find_gate(x:UOp) -> Optional[UOp]:
     if x.op is UOps.IF: return x
     return next((ret for s in x.src if (ret:=find_gate(s)) is not None), None)
-  if len(root.src) == 3 or (gate:=find_gate(root)) is None or gate.src[0] is not root.src[3]: return None
+  if len(root.src) == 3 or (gate:=find_gate(root)) is None or gate.src[0] is not root.src[3] \
+    and not ((sub_gate := find_gate(root.src[2])) is not None and sub_gate.src[0] is root.src[-1].src[0]):
+    return None
   return UOp(UOps.STORE, root.dtype, root.src[:3], root.arg)
 
+def merge_gates(sink:UOp) -> Optional[UOp]:
+  @functools.lru_cache(None)
+  def has_range(x:UOp) -> bool: return any(s2p.op is UOps.RANGE for s2p in x.sparents)
+  if len([x.op is UOps.STORE and len(x.src) == 4 for x in sink.src]) <= 1: return None
+  if_src_to_if_op: Dict[UOp, UOp] = {}
+  for x in sink.src:
+    if x.op is UOps.STORE and len(x.src) == 4 and has_range(x.src[2]) and x.src[2] not in x.src[-1].src:
+      if_to_update = if_src_to_if_op.get(x.src[-1].src[0], x.src[-1])
+      if_src_to_if_op[x.src[-1].src[0]] = UOp(UOps.IF, None, (if_to_update.src[0],) + if_to_update.src[1:] + (x.src[2],), if_to_update.arg)
```

Comment:
i dont remember if having an arg...

Reply:
I was just following suit with when the IF was getting constructed in `create_gate`, but `None` probably makes more sense, you're right. Will swap out